libs/stokesian_dynamics/sd.hpp

```diff
@@ -0,0 +1,1478 @@
+                tzm(i, 3) = 2.0 * (m(mesid(0, i), mesid(0, i), 1, 2) -
+                                   m(mesid(1, i), mesid(1, i), 1, 2));
+                tzm(i, 4) = m(mesid(0, i), mesid(0, i), 1, 1) -
+                            m(mesid(0, i), mesid(0, i), 2, 2) -
+                            (m(mesid(1, i), mesid(1, i), 1, 1) -
+                             m(mesid(1, i), mesid(1, i), 2, 2));
+
+            } else {
+
+                tzm(i, 0) = 2.0 * (m(mesid(0, i), mesid(1, i), 0, 0) -
+                                   m(mesid(0, i), mesid(1, i), 2, 2));
+                tzm(i, 1) = 4.0 * m(mesid(0, i), mesid(1, i), 0, 1);
+                tzm(i, 2) = 4.0 * m(mesid(0, i), mesid(1, i), 0, 2);
+                tzm(i, 3) = 4.0 * m(mesid(0, i), mesid(1, i), 1, 2);
+                tzm(i, 4) = 2.0 * (m(mesid(0, i), mesid(1, i), 1, 1) -
+                                   m(mesid(0, i), mesid(1, i), 2, 2));
+            }
+        }
+        //***************************************************************************c
+        //***************************************************************************c
+        //*** fill in upper half of m12 (=m11) and m22 (=m11).
+
+        for (std::size_t j = 0; j < 5; ++j) {
+
+            std::size_t j5 = j + 5;
+
+            for (std::size_t i = 0; i < j + 1; ++i) {
+
+                std::size_t i5 = i + 5;
+
+                tzm(i, j5) = visc12_3 * tzm(i, j);
+                tzm(i5, j5) = visc22_3 * tzm(i, j);
+            }
+        }
+        //*** fill in the lower half of m12.
+        for (std::size_t i = 0; i < 5; ++i) {
+
+            std::size_t i5 = i + 5;
+            for (std::size_t j = i + 1; j < 5; ++j) {
+
+                std::size_t j5 = j + 5;
+
+                tzm(j, i5) = visc12_3 * tzm(i, j5);
+            }
+        }
+
+        for (std::size_t i = 0; i < 5; ++i) {
+            for (std::size_t j = 0; j < 5; ++j) {
+                tzm(i, j) = visc11_3 * tzm(i, j);
+            }
+        }
+    }
+};
+
+template <typename T>
+struct thermalizer {
+    T sqrt_kT_Dt;
+    std::size_t offset;
+    std::size_t seed;
+#ifdef __CUDACC__
+    __device__
+#endif
+    T operator()(std::size_t index) {
+#ifdef __CUDACC__
+        uint4 rnd_ints = curand_Philox4x32_10(make_uint4(offset >> 32, seed >> 32, index >> 32, index),
+                                              make_uint2(offset, seed));
+        T rnd = _curand_uniform_double_hq(rnd_ints.w, rnd_ints.x);
+#else
+        // get two 64-bit random unsigned integers (of which only one is used)
+        RNG rng;
+        RNG::ctr_type c;
+        RNG::key_type k;
+        k[0] = seed;
+        c[0] = offset;
+        c[1] = index;
+        RNG::ctr_type rint = rng(c, k);
+        
+        // convert to uniform distribution
+        auto constexpr const max = std::numeric_limits<uint64_t>::max();
+        auto constexpr const fac = 1. / (max + 1.);
+        T rnd = rint[0] * fac;
```

Comment:
This seems to not produce the desired result on the Clang:6.0 image.

Reply:
This was copied from some other implementation. @mkuron knows where it came from.

Reply:
That is my own code. I have just never used this kind of RNG before and have no idea why it might fail on a certain compiler.

Reply:
This is copied from
https://github.com/espressomd/espresso/blob/434b9d75dc67026bcea1cd86456ec0fb4b6d1dab/src/utils/include/utils/uniform.hpp#L36-L40
Not sure why `+ 0.5 * fac` is missing at line 1192, but adding it doesn't change the outcome of the test in `clang:6.0` run locally, the obtained value is identical to the last decimal place. The `RNG` is `r123::Philox2x64`, other thermostats in the core use the `r123::Philox4x64` variant because of additional constraints on the RNG key.

Reply:
Ah yes, that is where lines 1190 and 1191 were taken from. It might be that `auto` fails to determine the right data type? 
Anyway, from my understanding, line 1192 scales the whole thing to the range [0,1]. The `+ 0.5 * fac` is a very small correction which should be included in line 1192.